Sphinx 2.0+ incompatibility

The docs build is not compatible with Sphinx 2.0. It was already mentioned in a PR recently, but I suppose it's better to have a separate issue for it, too.

Sphinx 1.8.5 works.

```
make html
# for testing
# @echo installed numpy 0a133311b5 matches git version 0a133311b5; exit 1
mkdir -p build/html build/doctrees
LANG=C sphinx-build -b html -WT --keep-going -d build/doctrees  -n source build/html 
Running Sphinx v2.1.2
1.18.dev0 1.18.0.dev0+0a13331

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/sphinx/config.py", line 361, in eval_config_file
    execfile_(filename, namespace)
  File "/usr/local/lib/python3.7/site-packages/sphinx/util/pycompat.py", line 86, in execfile_
    exec(code, _globals)
  File "/Users/bsipocz/munka/devel/numpy/doc/source/conf.py", line 374, in <module>
    class NumPyLexer(CLexer):
  File "/Users/bsipocz/munka/devel/numpy/doc/source/conf.py", line 377, in NumPyLexer
    tokens = copy.deepcopy(lexers['c'].tokens)
KeyError: 'c'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/sphinx/cmd/build.py", line 283, in build_main
    args.tags, args.verbosity, args.jobs, args.keep_going)
  File "/usr/local/lib/python3.7/site-packages/sphinx/application.py", line 209, in __init__
    self.config = Config.read(self.confdir, confoverrides or {}, self.tags)
  File "/usr/local/lib/python3.7/site-packages/sphinx/config.py", line 196, in read
    namespace = eval_config_file(filename, tags)
  File "/usr/local/lib/python3.7/site-packages/sphinx/config.py", line 371, in eval_config_file
    raise ConfigError(msg % traceback.format_exc())
sphinx.errors.ConfigError: There is a programmable error in your configuration file:

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/sphinx/config.py", line 361, in eval_config_file
    execfile_(filename, namespace)
  File "/usr/local/lib/python3.7/site-packages/sphinx/util/pycompat.py", line 86, in execfile_
    exec(code, _globals)
  File "/Users/bsipocz/munka/devel/numpy/doc/source/conf.py", line 374, in <module>
    class NumPyLexer(CLexer):
  File "/Users/bsipocz/munka/devel/numpy/doc/source/conf.py", line 377, in NumPyLexer
    tokens = copy.deepcopy(lexers['c'].tokens)
KeyError: 'c'


Configuration error:
There is a programmable error in your configuration file:

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/sphinx/config.py", line 361, in eval_config_file
    execfile_(filename, namespace)
  File "/usr/local/lib/python3.7/site-packages/sphinx/util/pycompat.py", line 86, in execfile_
    exec(code, _globals)
  File "/Users/bsipocz/munka/devel/numpy/doc/source/conf.py", line 374, in <module>
    class NumPyLexer(CLexer):
  File "/Users/bsipocz/munka/devel/numpy/doc/source/conf.py", line 377, in NumPyLexer
    tokens = copy.deepcopy(lexers['c'].tokens)
KeyError: 'c'

make: *** [html-build] Error 2
```

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -21,7 +21,7 @@ jobs:
             python3 -m venv venv
             ln -s $(which python3) venv/bin/python3.6
             . venv/bin/activate
-            pip install cython sphinx==1.8.5 matplotlib ipython
+            pip install cython sphinx==2.1.2 matplotlib ipython
             sudo apt-get update
             sudo apt-get install -y graphviz texlive-fonts-recommended texlive-latex-recommended texlive-latex-extra texlive-generic-extra latexmk texlive-xetex
 

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -21,7 +21,7 @@ jobs:
             python3 -m venv venv
             ln -s $(which python3) venv/bin/python3.6
             . venv/bin/activate
-            pip install cython sphinx==2.1.2 matplotlib ipython
+            pip install cython sphinx==2.2.0 matplotlib ipython
             sudo apt-get update
             sudo apt-get install -y graphviz texlive-fonts-recommended texlive-latex-recommended texlive-latex-extra texlive-generic-extra latexmk texlive-xetex
 

--- a/azure-pipelines.yml
+++ b/azure-pipelines.yml
@@ -88,7 +88,7 @@ jobs:
     displayName: 'Install tools'
   - script: |
       python -m pip install -r test_requirements.txt
-      python -m pip install vulture docutils sphinx==1.8.5 numpydoc
+      python -m pip install vulture docutils sphinx==2.1.2 numpydoc
     displayName: 'Install dependencies; some are optional to avoid test skips'
   - script: /bin/bash -c "! vulture . --min-confidence 100 --exclude doc/,numpy/distutils/ | grep 'unreachable'"
     displayName: 'Check for unreachable code paths in Python modules'

--- a/azure-pipelines.yml
+++ b/azure-pipelines.yml
@@ -88,7 +88,7 @@ jobs:
     displayName: 'Install tools'
   - script: |
       python -m pip install -r test_requirements.txt
-      python -m pip install vulture docutils sphinx==2.1.2 numpydoc
+      python -m pip install vulture docutils sphinx==2.2.0 numpydoc
     displayName: 'Install dependencies; some are optional to avoid test skips'
   - script: /bin/bash -c "! vulture . --min-confidence 100 --exclude doc/,numpy/distutils/ | grep 'unreachable'"
     displayName: 'Check for unreachable code paths in Python modules'

--- a/doc/source/_templates/autosummary/base.rst
+++ b/doc/source/_templates/autosummary/base.rst
@@ -1,0 +1,14 @@
+{% if objtype == 'property' %}
+:orphan:
+{% endif %}
+
+{{ fullname | escape | underline}}
+
+.. currentmodule:: {{ module }}
+
+{% if objtype == 'property' %}
+property
+{% endif %}
+
+.. auto{{ objtype }}:: {{ objname }}
+

--- a/doc/source/conf.py
+++ b/doc/source/conf.py
@@ -4,7 +4,7 @@ from __future__ import division, absolute_import, print_function
 import sys, os, re
 
 # Minimum version, enforced by sphinx
-needs_sphinx = '2.1.2'
+needs_sphinx = '2.2.0'
 
 # -----------------------------------------------------------------------------
 # General configuration

--- a/doc/source/conf.py
+++ b/doc/source/conf.py
@@ -3,12 +3,8 @@ from __future__ import division, absolute_import, print_function
 
 import sys, os, re
 
-# Check Sphinx version
-import sphinx
-if sphinx.__version__ < "1.2.1":
-    raise RuntimeError("Sphinx 1.2.1 or newer required")
-
-needs_sphinx = '1.0'
+# Minimum version, enforced by sphinx
+needs_sphinx = '2.1.2'
 
 # -----------------------------------------------------------------------------
 # General configuration
@@ -31,19 +27,18 @@ extensions = [
     'matplotlib.sphinxext.plot_directive',
     'IPython.sphinxext.ipython_console_highlighting',
     'IPython.sphinxext.ipython_directive',
+    'sphinx.ext.imgmath',
 ]
 
-if sphinx.__version__ >= "1.4":
-    extensions.append('sphinx.ext.imgmath')
-    imgmath_image_format = 'svg'
-else:
-    extensions.append('sphinx.ext.pngmath')
+imgmath_image_format = 'svg'
 
 # Add any paths that contain templates here, relative to this directory.
 templates_path = ['_templates']
 
 # The suffix of source filenames.
 source_suffix = '.rst'
+
+master_doc = 'contents'
 
 # General substitutions.
 project = 'NumPy'
@@ -93,6 +88,7 @@ pygments_style = 'sphinx'
 def setup(app):
     # add a config value for `ifconfig` directives
     app.add_config_value('python_version_major', str(sys.version_info.major), 'env')
+    app.add_lexer('NumPyC', NumPyLexer(stripnl=False))
 
 # -----------------------------------------------------------------------------
 # HTML output
@@ -176,6 +172,10 @@ latex_documents = [
 # For "manual" documents, if this is true, then toplevel headings are parts,
 # not chapters.
 #latex_use_parts = False
+
+latex_elements = {
+    'fontenc': r'\usepackage[LGR,T1]{fontenc}'
+}
 
 # Additional stuff for the LaTeX preamble.
 latex_preamble = r'''
@@ -368,18 +368,15 @@ def linkcode_resolve(domain, info):
 
 from pygments.lexers import CLexer
 from pygments import token
-from sphinx.highlighting import lexers
 import copy
 
 class NumPyLexer(CLexer):
     name = 'NUMPYLEXER'
 
-    tokens = copy.deepcopy(lexers['c'].tokens)
+    tokens = copy.deepcopy(CLexer.tokens)
     # Extend the regex for valid identifiers with @
     for k, val in tokens.items():
         for i, v in enumerate(val):
             if isinstance(v, tuple):
                 if isinstance(v[0], str):
                     val[i] =  (v[0].replace('a-zA-Z', 'a-zA-Z@'),) + v[1:]
-
-lexers['NumPyC'] = NumPyLexer(stripnl=False)

